File: LeetCode/211.Design_Add_and_Search_Words_Data_Structure/susan/211.DesignAdd_and_Search_Words_Data_Structure.java
class WordDictionary {

    TrieNode root;

    public WordDictionary() {
        root = new TrieNode();
    }
    
    public void addWord(String word) {
        TrieNode node = root;
        for(char c : word.toCharArray()){
            if(node.children[c-'a']== null){
                node.children[c-'a'] = new TrieNode();
            }
            node = node.children[c-'a'];
        }
        node.isWork = true;
    }
    
    public boolean search(String word) {
        return helper(word,0,root);
    }

    public boolean helper(String word, int position, TrieNode node){
         if(position == word.length()){
             return node.isWork;
         }
         char c = word.charAt(position);
         if( c != '.'){
             return node.children[c-'a'] != null && helper(word, position+1,node.children[c-'a']);
         }

         for(int i = 0; i< 26; i++){
             if(node.children[i] != null && helper(word,position+1,node.children[i])){
                 return true;
             }
         }

         return false;
    }
}
class TrieNode {
    TrieNode[] children = new TrieNode[26];
    boolean isWork;
}

File: LeetCode/211.Design_Add_and_Search_Words_Data_Structure/211.Design_Add_and_Search_Words_Data_Structure.py
#!/usr/bin/env python3


class TrieNode:

    def __init__(self):
        self.children = {}  # key:str, val: TrieNode
        self.isEnd = False


class WordDictionary:

    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = TrieNode()
            node = node.children[c]
        node.isEnd = True

    def search(self, word: str) -> bool:
        return self._helper(word, self.root)

    def _helper(self, word, node0):
        node = node0
        for i, c in enumerate(word):
            if c in node.children:
                node = node.children[c]
            elif c == ".":
                if len(node.children.keys()) == 0:
                    return False
                for key in node.children:
                    part = word[i + 1 :]
                    if self._helper(key + part, node):
                        return True
                return False
            else:
                return False

        return node.isEnd


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)
